File: Note/app/src/main/java/com/zukron/note/fragment/DefaultNoteFragment.java
package com.zukron.note.fragment;

import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.google.android.material.textfield.TextInputEditText;
import com.zukron.note.R;
import com.zukron.note.model.DefaultNote;
import com.zukron.note.model.Note;
import com.zukron.note.model.dao.DefaultNoteDao;

/**
 * Project name is Note
 * Created by Zukron Alviandy R on 8/8/2020
 */
public class DefaultNoteFragment extends Fragment {
    public static final String TAG = DefaultNoteFragment.class.getSimpleName();
    private static final String EXTRA_NOTE = "note";
    private TextInputEditText inputBodyNote, inputTitleNote;
    private Integer id = null;

    public DefaultNoteFragment() {
    }

    public static DefaultNoteFragment newInstance(Note note) {
        DefaultNoteFragment defaultNoteFragment = new DefaultNoteFragment();
        Bundle bundle = new Bundle();
        bundle.putParcelable(EXTRA_NOTE, note);

        defaultNoteFragment.setArguments(bundle);
        return defaultNoteFragment;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.default_note_fragment, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        inputTitleNote = view.findViewById(R.id.input_title_note);
        inputBodyNote = view.findViewById(R.id.input_body_note);
    }

    @Override
    public void onStart() {
        super.onStart();

        DefaultNoteDao defaultNoteDao = new DefaultNoteDao(getContext());

        // jika user ingin mengupdate
        if (getArguments() != null) {
            Note note = getArguments().getParcelable(EXTRA_NOTE);

            if (note != null) {
                DefaultNote defaultNote = defaultNoteDao.get(note.getId());
                id = defaultNote.getId();
                inputTitleNote.setText(defaultNote.getTitle());
                inputBodyNote.setText(defaultNote.getBody());
            }
        }
    }

    public DefaultNote getNote() {
        DefaultNote defaultNote = null;

        if (validated()) {
            if (inputBodyNote.getText() != null && inputTitleNote.getText() != null) {
                String title = inputTitleNote.getText().toString().trim();
                String body = inputBodyNote.getText().toString().trim();

                defaultNote = new DefaultNote(id, title, body);
            }
        } else {
            Toast.makeText(getContext(), "Judul atau isi tidak boleh kosong", Toast.LENGTH_SHORT).show();
        }

        return defaultNote;
    }

    private boolean validated() {
        boolean valid = true;

        assert inputTitleNote.getText() != null;
        if (TextUtils.isEmpty(inputTitleNote.getText().toString())) {
            valid = false;
        }

        assert inputBodyNote.getText() != null;
        if (TextUtils.isEmpty(inputBodyNote.getText().toString())) {
            valid = false;
        }

        return valid;
    }
}

File: Note/app/src/main/java/com/zukron/note/ui/fragment/DefaultNoteFragment.kt
package com.zukron.note.ui.fragment

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import com.zukron.note.R
import com.zukron.note.model.DefaultNote
import com.zukron.note.ui.viewmodel.NoteViewModel
import com.zukron.note.util.Utilities.toEditable
import kotlinx.android.synthetic.main.default_note_fragment.*

/**
 * Project name is Note
 * Created by Zukron Alviandy R on 9/5/2020
 * Contact me if any issues on zukronalviandy@gmail.com
 */
class DefaultNoteFragment : Fragment() {
    companion object {
        const val TAG: String = "DefaultNoteFragment"
        private const val BUNDLE_NOTE_ID: String = "bundle_note_id"

        fun newInstance(id: Long): DefaultNoteFragment {
            val defaultNoteFragment = DefaultNoteFragment()
            val bundle = Bundle()
            bundle.putLong(BUNDLE_NOTE_ID, id)

            defaultNoteFragment.arguments = bundle
            return defaultNoteFragment
        }
    }

    private lateinit var noteViewModel: NoteViewModel
    private var _defaultNote: DefaultNote? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.default_note_fragment, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        // note id from Note entity
        val noteId = arguments?.getLong(BUNDLE_NOTE_ID, 0) ?: 0

        // view model
        noteViewModel = ViewModelProvider(this, ViewModelProvider
                .AndroidViewModelFactory(activity!!.application))
                .get(NoteViewModel::class.java)

        noteViewModel.setNoteId(noteId)
        noteViewModel.defaultNote.observe(this) {
            if (it != null) {
                input_title.text = it.title.toEditable()
                input_body.text = it.body.toEditable()
                _defaultNote = it
            }
        }
    }

    fun getNote(): DefaultNote? {
        var defaultNote: DefaultNote? = _defaultNote

        if (validate()) {
            val title = input_title.text.toString().trim()
            val body = input_body.text.toString().trim()

            // insert
            if (defaultNote == null) {
                defaultNote = DefaultNote(body)
            }

            //update
            defaultNote.body = body
            defaultNote.title = title
        } else {
            Toast.makeText(context, "Judul atau isi tidak boleh kosong", Toast.LENGTH_SHORT).show()
        }

        return defaultNote
    }

    private fun validate(): Boolean {
        return when {
            input_title.text!!.isBlank() -> false
            input_body.text!!.isBlank() -> false
            else -> true
        }
    }
}